File: dart/packages/aoc2019/lib/src/day13.dart
// https://adventofcode.com/2019/day/13

import 'package:aoc/aoc.dart';
import 'package:aoc/util/vec.dart';
import 'package:collection/collection.dart';

import 'intcode.dart';

main() => Day13().solve();

const int empty = 0;
const int wall = 1;
const int block = 2;
const int paddle = 3;
const int ball = 4;

class Day13 extends AdventDay {
  Day13() : super(2019, 13);

  @override
  dynamic part1(String input) {
    final machine = Intcode.from(program: input);
    while (!machine.execute()) {}
    return machine.output.slices(3).where((p) => p.last == block).length;
  }

  @override
  dynamic part2(String input) {
    final machine = Intcode.from(program: input);
    int score = 0;
    Vec ballPos = Vec.zero;
    Vec paddlePos = Vec.zero;

    machine[0] = 2;
    while (!machine.execute() || machine.output.isNotEmpty) {
      machine.output.slices(3).map((p) => p.toList()).forEach((command) {
        if (command[0] == -1 && command[1] == 0) {
          score = command[2];
        } else {
          switch (command[2]) {
            case ball:
              ballPos = Vec(command[0], command[1]);
              break;
            case paddle:
              paddlePos = Vec(command[0], command[1]);
              break;
          }
        }
      });
      machine.output.clear();
      // Simplistic AI: just move the paddle in the direction of the ball.
      machine.input.add(ballPos.xInt.compareTo(paddlePos.xInt));
    }
    return score;
  }
}

File: clojure/src/aoc2019/day13.clj
;; https://adventofcode.com/2019/day/13
 (ns aoc2019.day13
   (:require
    [aoc.day :as d]
    [aoc.util.collection :as c]
    [aoc2019.intcode :as ic]))

(defn input [] (d/day-input 2019 13))

(def ^:private block  2)
(def ^:private paddle 3)
(def ^:private ball   4)

(defn- update-state [state [x y value]]
  (cond
    (= [x y] [-1 0]) (assoc state :score value)
    (= value paddle) (assoc state :paddle [x y])
    (= value ball)   (assoc state :ball [x y])
    :else state))

(defn- joystick-input [state]
  ;; Simple strategy: move the paddle in the direction of the ball
  (let [ball-x (first (:ball state))
        paddle-x (first (:paddle state))]
    (compare ball-x paddle-x)))

(defn part1 [input]
  (->> (ic/parse-program input)
       ic/run
       :output
       (partition 3)
       (map last)
       (c/count-where #{block})))

(defn part2 [input]
  (let [game (-> (ic/parse-program input)
                 ic/init-machine
                 (assoc-in [:mem 0] 2)
                 ic/execute)]
    (loop [{:keys [status output] :as game} game, state {}]
      (if (and (= :halted status) (empty? output))
        (:score state)
        (let [state' (reduce update-state state (partition 3 output))
              joystick (joystick-input state')]
          (recur (ic/execute (ic/update-io game [joystick])) state'))))))
